src/node_sets.cpp

```diff
@@ -0,0 +1,317 @@
+
+template <typename T>
+std::string toString(const std::string& key, const std::vector<T>& values) {
+    return fmt::format(R"("{}": [{}])", key, fmt::join(values, ", "));
+}
+
+template <>
+std::string toString(const std::string& key, const std::vector<std::string>& values) {
+    // strings need to be wrapped in quotes
+    return fmt::format(R"("{}": ["{}"])", key, fmt::join(values, "\", \""));
+}
+
+template <typename T>
+class NodeSetBasicRule: public NodeSetRule
+{
+  public:
+    NodeSetBasicRule(std::string attribute, std::vector<T>&& values)
+        : attribute_(std::move(attribute))
+        , values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& v : values_) {
+            ret = ret | np.matchAttributeValues(attribute_, v);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString(attribute_, values_);
+    }
+
+  private:
+    std::string attribute_;
+    std::vector<T> values_;
+};
+
+class NodeSetBasicPopulation: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicPopulation(std::string name)
+        : name_(std::move(name)) {}
+
+    Selection materialize(const NodeSets& /* unused */, const NodePopulation& np) const final {
+        if (name_ == np.name()) {
+            return np.selectAll();
+        }
+
+        return Selection{{}};
+    }
+
+    std::string toJSON() const final {
+        const std::vector<std::string> values = {name_};
+        return toString("population", values);
+    }
+
+  private:
+    std::string name_;
+};
+
+class NodeSetBasicNodeIds: public NodeSetRule
+{
+  public:
+    explicit NodeSetBasicNodeIds(Selection::Values&& values)
+        : values_(values) {}
+
+    Selection materialize(const NodeSets& /* unused */,
+                          const NodePopulation& /* unused */) const final {
+        return Selection::fromValues(values_.begin(), values_.end());
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", values_);
+    }
+
+  private:
+    Selection::Values values_;
+};
+
+using CompoundTargets = std::vector<std::string>;
+class NodeSetCompoundRule: public NodeSetRule
+{
+  public:
+    NodeSetCompoundRule(std::string attribute, const CompoundTargets& targets)
+        : attribute_(std::move(attribute))
+        , targets_(targets) {}
+
+    Selection materialize(const NodeSets& ns, const NodePopulation& np) const final {
+        Selection ret{{}};
+        for (const auto& target : targets_) {
+            ret = ret | ns.materialize(target, np);
+        }
+        return ret;
+    }
+
+    std::string toJSON() const final {
+        return toString("node_ids", targets_);
+    }
+
+  private:
+    std::string attribute_;
+    CompoundTargets targets_;
+};
+
+int64_t get_integer_or_throw(const json& el) {
+    auto v = el.get<double>();
+    if (std::floor(v) != v) {
```

Comment:
std::numeric_limits<double>::epsilon ?

Reply:
Why? I want to make sure that it's an integer, not that it's close to an integer.

Reply:
ah ok when I first read the function I though you wanted something close to an integer. 
Then don t you need abs() if the integer is a negative number ?

Reply:
Negative numbers are allowed; it's not just `node_id` that can be tested against.

Reply:
You are using this here :

https://github.com/BlueBrain/libsonata/pull/107/files#diff-2eb0c657a8cf02cb85ca88ac45b71f6a3746c46ab114981bea14e605449a3325R138
So the NodeSetRules with an int attribute will also go inside this function.

So my question was : 
is `std::floor(v)  != v)` always safe for a negative number ? 
That is, is it possible `el.get<double> ` returns for some reason a number such as : -9.0000000000000000000000000001 and then the floor will be -10 instead of -9 ? 

Shouldn't we use : 
```
std::floor(std::abs(v))  != v
```
for safety ?

Reply:
> el.get<double> returns for some reason a number such as : -9.0000000000000000000000000001

All integers should be representable by float/doubles up to (about) 2**sizeof_mantissa, so `el.get<double>` shouldn't returning anything that requires `floor`ing, unless they aren't integers, regardless of sign.

Reply:
yeah indeed. 
All good then